File: 백준/Gold/7576. 토마토/토마토.js
const fs = require('fs');
const input = fs
    .readFileSync(0, 'utf-8')
    .toString()
    .trim()
    .split('\n')
    .map(line => line.split(' ').map(Number));
const [M, N] = input[0];
const box = input.slice(1);
const queue = [];
let total = 0;
let riped = 0;

for (let r = 0; r < N; r++) {
    for (let c = 0; c < M; c++) {
        if (box[r][c] === 0) total++;
        if (box[r][c] === 1) {
            total++;
            riped++;
            
            queue.push([r, c]);
        }
    }
}

if (total === riped) {
    console.log(0);
    process.exit();
}

const dr = [-1, 0, 1, 0];
const dc = [0, 1, 0, -1];
let index = 0;
let result;

while (index < queue.length) {
    const [cr, cc] = queue[index++];
    result = box[cr][cc];
    
    for (let i = 0; i < 4; i++) {
        const nr = cr + dr[i];
        const nc = cc + dc[i];
        
        if (nr < 0 || nr >= N) continue;
        if (nc < 0 || nc >= M) continue;
        if (box[nr][nc] !== 0) continue;
        
        riped++;
        box[nr][nc] = result + 1;
        queue.push([nr, nc]);
    }
}

console.log(riped < total ? -1 : result - 1);

File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque
import sys

M, N = map(int,input().split())
box = [list(map(int, input().split())) for _ in range(N)]
count_unriped = 0
queue_riped = deque()
day = 0
di = [1, 0, -1, 0]
dj = [0, 1, 0, -1]

for i in range(N):
    for j in range(M):
        if box[i][j] == 0:
            count_unriped += 1
        if box[i][j] == 1:
            queue_riped.append([i, j])

if (count_unriped == 0):
    print(0)
    sys.exit()

while count_unriped > 0 and len(queue_riped) > 0:
    ci, cj = queue_riped.popleft()

    for k in range(4):
        ni = ci + di[k]
        nj = cj + dj[k]

        if (ni >= 0 and nj >= 0 and ni < N and nj < M
           and box[ni][nj] == 0):
            day = box[ci][cj] + 1
            box[ni][nj] = day
            count_unriped -= 1
            queue_riped.append([ni, nj])

print(day-1 if count_unriped == 0 else -1)